### ISSUE TITLE
apply gaussian blur

### ISSUE THREAD

**Issue by shuaigezhu** [shuaigezhu/star_galaxy_classification#30]

**Comment by shuaigezhu** [comment]
Hi Kwang. 
Do you have any advice about how should I select 'label-pixels' and 'unlabel-pixels' (unlabeled pixels are meaning of 25% for each class) after I apply gaussian blur? In the past, I just need to split them as 25% points and non-25% points. The reason I ask this question is 'label-pixels' and 'unlabel-pixels' are highly related to the weights in loss function.

**Comment by kmyi** [comment]
You can threshold with a certain value for each label and then use that instead of selecting 25% and non-25%. 
--
Kwang Moo YI
Assistant Professor
Dept. of Computer Science
University of Victoria
+1 (250) 472 5837

**Comment by sfabbro** [comment]
Since you are doing regression now rather than classification, do you need to split your pixels? Why not simply applying your Gaussian blur over the whole label maps after you set your  background pixels as 0.25?

**Comment by kmyi** [comment]
I would imagine you still need to balance training samples :) Regression also needs some balancingg

Message written with iPhone. 
Please excuse any brievity and typos.
—
Kwang Moo YI
Assistant Professor
Dept. of Computer Science
University of Victoria
+1 (250) 475 5837

**Comment by shuaigezhu** [comment]
> Since you are doing regression now rather than classification, do you need to split your pixels? Why not simply applying your Gaussian blur over the whole label maps after you set your background pixels as 0.25?

Yes, I have applied Gaussian blur over the whole label maps, but then I need to decide the weights of unlabeled pixels and labeled pixels. The reason I split it up is I don't want the unlabeled pixels with actual objects there, to have the same weight as the labeled pixels.

**Comment by sfabbro** [comment]
You can do something like this:
1 - Make your probability label maps P, with 0.25 on the background pixels
2 - Make a probability variance map with a large error on background pixels
(>0.5) and small error on labels (0.01). Then you do;
       Var[P] = 0.01^2  on labels, Var[P] = 0.5^2 on unlabelled
3 - Make a Gaussian kernel k. Apply k to P, and apply kernel k² to Var[P]
4 - Apply weights on all pixels as weight = 1/Var[P], replacing the P=0
with P=1e-6 to avoid NaN
6 - Use the MSE weighted loss sum( (P_model - P_label)^2 * weight )

**Comment by kmyi** [comment]
Please post the final outcome, example images before you close the issues. Can you do that for the rest as well?

**Comment by shuaigezhu** [comment]
yes, of course

Kwang Moo Yi <notifications@github.com> 于2020年4月13日周一 下午12:30写道：

**Comment by shuaigezhu** [comment]
> Please post the final outcome, example images before you close the issues. Can you do that for the rest as well?

In this issue, for example, the gaussian blur would make central pixel with a value of 1 spread to 9 nearby pixels with the value of [0.33024799 0.33483236 0.33024799 0.33483236 0.33967865 0.33483236
 0.33024799 0.33483236 0.33024799]  
I uploaded the evaluation of final outcome by the confusion matrix for 6 types of model in issue **#35 Confusion Matrix**, and anything else I should post?

**Comment by shuaigezhu** [comment]
![Screen Shot 2020-08-17 at 1 48 22 PM](https://user-images.githubusercontent.com/47727949/90443033-6d300f00-e090-11ea-9870-e0952ccc7cd5.png)
The yellow point is the pixel with prob = 1

![Screen Shot 2020-08-17 at 1 48 29 PM](https://user-images.githubusercontent.com/47727949/90443085-83d66600-e090-11ea-9a5e-4889b1165f91.png)
After applied gaussian blur
